### PR TITLE
Consolidate HDCED4 and HDCED into Single Calculator

### PR DESCRIPTION
**Pull request by joelstobart-moj**
- Removed all instances of HDCED4PLUS from CalculationRule.kt and ComparisonType.kt to consolidate HDCED4 and HDCED rules. 
- Updated SentenceCalculation.kt to reflect the unified handling of HDCED rules. 
- Simplified the comparison logic by removing the separate ESTABLISHMENT_HDCED4PLUS comparison type. 
- Refactored methods to work with the single HDCED calculator, ensuring consistent calculation logic across both previous implementations. 
- Improved readability and maintainability by streamlining calculation rules and reducing redundant logic.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt [hunk 1]
```diff
@@ -105,38 +168,59 @@ class HdcedCalculator(val hdcedConfiguration: HdcedConfiguration) {
...
+  private fun getDeductedDays(sentenceCalculation: SentenceCalculation) =
+    sentenceCalculation.calculatedTotalDeductedDays.toLong()
+
   private fun calculateHdcedMinimumCustodialPeriod(
+    sentence: CalculableSentence,
     sentenceCalculation: SentenceCalculation,
-    sentenceDate: LocalDate,
     parentRule: CalculationRule,
-    addedDays: Long,
+    params: Hdced4Params,
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt:182]: Perhaps next iteration is to rename this?

**Review by srowlands-moj — COMMENTED** [review]
Looking pretty good otherwise, I did get some slightly unexpected test results when I did something similar, hopefully you don't get the same, in some tests changing what was the HDCED4PLUS test result to be the HDCED (When they were different) a few didn't tie up.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt (hunk 1 above)
Comment (joelstobart-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt:182]: No - I should fix that too!

Comment (joelstobart-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt]: Done - I think

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/enumerations/ComparisonType.kt [hunk 2]
```diff
@@ -2,10 +2,9 @@ package uk.gov.justice.digital.hmpps.calculatereleasedatesapi.enumerations
...
 enum class ComparisonType {
   ESTABLISHMENT_FULL,
-  ESTABLISHMENT_HDCED4PLUS,
   MANUAL,
 }
 
-public fun manualComparisonTypes() = setOf(ComparisonType.MANUAL)
+fun manualComparisonTypes() = setOf(ComparisonType.MANUAL)
 
-public fun nonManualComparisonTypes() = setOf(ComparisonType.ESTABLISHMENT_FULL, ComparisonType.ESTABLISHMENT_HDCED4PLUS)
```
[the remark this answers is not in the document]
Comment (joelstobart-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/enumerations/ComparisonType.kt:11]: I've added a script to delete historic HDCED4 comparisons

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt [hunk 3]
```diff
@@ -1,102 +1,165 @@
...
 import java.time.LocalDate
-import java.time.temporal.ChronoUnit
 import kotlin.math.ceil
 import kotlin.math.max
 
 @Service
-class HdcedCalculator(val hdcedConfiguration: HdcedConfiguration) {
+class HdcedCalculator(
+  val hdcedConfiguration: HdcedConfiguration,
+) {
```
[the remark this answers is not in the document]
Comment (joelstobart-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/HdcedCalculator.kt:23]: At least we're hopefully getting this sorted now :)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingExtractionService.kt [hunk 4]
```diff
@@ -190,15 +169,17 @@ class BookingExtractionService(
...
         booking.sentenceGroups,
         sentences,
         mostRecentSentencesByReleaseDate,
       )
+     */
 
-    val latestHDC4PLUSAndBreakdown =
-      extractManyHomeDetentionCurfew4PlusEligibilityDate(sentences, mostRecentSentencesByReleaseDate)
+    val latestHDCEDAndBreakdown =
+      hdcedExtractionService.extractManyHomeDetentionCurfew4PlusEligibilityDate(sentences, mostRecentSentencesByReleaseDate)
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingExtractionService.kt:182]: Still refs hdc4

**Review by srowlands-moj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingExtractionService.kt (hunk 4 above)
Comment (joelstobart-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingExtractionService.kt:182]: Yep - going to resolve :(

**Review by srowlands-moj — APPROVED** [review]
LGTM

**Pull request merged**